Add endpoint for uploading books

andersnylund opened the pull request:
Was first thinking about doing the insertion of books through some scripts or something like that, but found out that NestJS provides file upload. With help of that it is now possible to upload a CSV file in a POST request where a form-data field `file` includes the CSV file. The file has to have a header row where names of the rows must match with the book entity column names.

Commit:
First draft of uploading books

Commit:
Add validation

andersnylund commented:
The goodreads-best-books.zip dataset is basically useless, as the ISBNs in it have been transformed at some point into `9.78038E+12`. Do you @mansour-ahmed happen to have access to the original source of the data?

Otherwise, uploading the CSVs combined with  `class-validator` seems promising 👍🏻

mansour-ahmed commented:
> Was first thinking about doing the insertion of books through some scripts or something like that, but found out that NestJS provides file upload. With help of that it is now possible to upload a CSV file in a POST request where a form-data field `file` includes the CSV file. The file has to have a header row where names of the rows must match with the book entity column names.

This sounds very good!

mansour-ahmed commented:
> The goodreads-best-books.zip dataset is basically useless, as the ISBNs in it have been transformed at some point into `9.78038E+12`. Do you @mansour-ahmed happen to have access to the original source of the data?
> 
> Otherwise, uploading the CSVs combined with `class-validator` seems promising 👍🏻

Yeah I noticed the same thing, I will try to find a better copy but for now let's use the other csv files! 👍🏼

mansour-ahmed commented on backend/README.md:
````diff
@@ -1,75 +1,50 @@
+
+Insert following environment variables into `.env`
+
+```env
````
Would a .env.example file be any better here?

mansour-ahmed commented on backend/src/app.module.ts:
```diff
@@ -1,12 +1,30 @@
+
+const { DB_PORT, DB_USERNAME, DB_PASSWORD, DB_DATABASE, DB_HOST } = process.env;
+
+const ormOptions: TypeOrmModuleOptions = {
```
Suggestion: Let's move these to an own configuration file e.g. `configuration/configuration.ts`

mansour-ahmed commented on backend/src/books/books.service.ts:
```diff
@@ -1,13 +1,60 @@
+  return readableStream;
+};
+
+const validateBooks = async (
```
Nestjs does automatically the validation if you provde the `Dto` definition to the controller.

andersnylund replied:
Yes, but it doesn't work in this case because we are sending a file in form-data and not JSON in the body

mansour-ahmed replied:
Fair enough!

mansour-ahmed commented on backend/src/books/books.service.ts:
```diff
@@ -18,4 +65,23 @@ export class BooksService {
+
+  uploadBooks(file: { buffer: Buffer }): void {
+    const stream = bufferToStream(file.buffer);
+    Papa.parse(stream, {
```
nice! 👌🏽

Commit:
Small fixes

- Move db configuration to its own file
- Add .env.example file

Commit:
Add tests

andersnylund commented:
@mansour-ahmed I added tests for the books controller and service. I also introduced a sqlite in-memory database that enables including entity operations in tests, and the repositories are not required to be mocked out

mansour-ahmed commented on backend/README.md:
```diff
@@ -1,75 +1,50 @@
 
 # e2e tests
-$ npm run test:e2e
+$ npm test:e2e
```
Should this be yarn instead?

```suggestion
$ yarn test:e2e
```

mansour-ahmed commented on backend/README.md:
```diff
@@ -1,75 +1,50 @@
 
 # test coverage
-$ npm run test:cov
+$ npm test:cov
```
Same question here about `npm`.

mansour-ahmed commented on backend/src/db.ts:
```diff
@@ -0,0 +1,19 @@
+
+const { DB_PORT, DB_USERNAME, DB_PASSWORD, DB_DATABASE, DB_HOST } = process.env;
+
+export const ormOptions: TypeOrmModuleOptions = {
```
I would suggest moving these into an own configuration file. e.g. `src/configuration/configuraiton.ts`

mansour-ahmed replied:
Ops, I just realized I mentioned this earlier, sorry.

I might have miscommunicated here, but I meant having a generic configurations file. This will do for now though! 👍🏼

mansour-ahmed commented on backend/src/db.ts:
```diff
@@ -0,0 +1,19 @@
+  username: DB_USERNAME,
+  password: DB_PASSWORD,
+  database: DB_DATABASE,
+  entities: [Book],
```
Suggestion, add the entities on the module level. e.g. inside BookModule:

```
@Module({
  controllers: [BooksController],
  imports: [HttpModule, TypeOrmModule.forFeature([BookEntity])],
  providers: [BooksService],
})
export class BooksModule {}
```

mansour-ahmed commented on backend/src/main.ts:
```diff
@@ -4,6 +4,7 @@ import { AppModule } from './app.module';
 
 async function bootstrap(): Promise<void> {
   const app = await NestFactory.create(AppModule);
+  app.setGlobalPrefix('api');
```
nice!

mansour-ahmed commented on backend/src/books/books.service.ts:
```diff
@@ -1,13 +1,60 @@
+import { BookDto } from './book.dto';
+
+/** https://stackoverflow.com/a/54136803 */
+const bufferToStream = (buffer: Buffer): Readable => {
```
Out of curiosity, would this work if its inside the class and written as a class function? 🤔

mansour-ahmed commented on backend/src/books/books.service.ts:
```diff
@@ -1,13 +1,60 @@
+  ) {}
 
-  async getBooks(query: string): Promise<Book[]> {
+  async getBooks(query: string): Promise<BookType[]> {
```
It is just a personal opinion but I find it easier to deal with observables. On the other hand this is not really a stream of data per say. No action point is needed here though in case you would like to keep the promises! 😊

mansour-ahmed commented on backend/src/books/books.service.ts:
```diff
@@ -18,4 +65,30 @@ export class BooksService {
+              errors,
+            });
+          } catch (e) {
+            reject(e);
```
I would suggest using Nestjs's Logger to log errors => `Logger.error()` , it can be imported from the `@nestjs/common` module

mansour-ahmed commented:
> @mansour-ahmed I added tests for the books controller and service. I also introduced a sqlite in-memory database that enables including entity operations in tests, and the repositories are not required to be mocked out

Looking good, the sqlite will come in handy as well! Later on we could look into integration tests! 👍🏼

andersnylund commented:
This PR has become deprecated, but I think we should still keep it open as a reference for how to write tests etc.

mansour-ahmed commented:
> This PR has become deprecated, but I think we should still keep it open as a reference for how to write tests etc.

Agreed. Should we perhaps close the PR though? It will still be available for our reference even if it is closed.
